docs/interpreter_design_doc.md

```diff
@@ -0,0 +1,100 @@
+# Interpreter Design
+
+
+## Class Hierarchy
+
+Stablehlo programs are computations over Tensors (n-dimensional arrays), hence
+most of the input/output data values for an op are Tensors with the exception of
+constant
+[Attribute](https://docs.google.com/document/d/1GAAK_yHk7q2cZdJ3zBhsiNcrqzfpyjwQ-oWLD7q__Ss/edit#heading=h.5nem09i7p9va)
+values of primitive types (e.g., Int, float etc.).
+
+In the current design, the class `Tensor` implements Tensor data. The underlying
+storage class for `Tensor` objects, `detail:Buffer`, stores the type of the
+Tensor along with a contiguous byte array representing its data.
+
+Individual elements of a Tensor are represented using `Element` class which uses
+`mlir::Attribute` for storage.
+
+`Tensor`
+class has the following APIs to interact with it's individual elements:
+  - `Element Tensor::get(int64_t index)`: To extract an individual tensor
+  element at index `index` as `Element` object.
+  - `void Tensor::set(int64_t index, Element element);`: To insert an `Element`
+  object `element` into a Tensor at index `index`.
```

Comment:
"For the skeleton of the interpreter, we chose a linearized API for accessing tensor elements to simplify the implementation. This is sufficient for elementwise ops, and in the future we're planning to expand this to cover more complicated op".